File: backend/app/routers/admin/invoices.py
from datetime import datetime
from fastapi import APIRouter, Depends, HTTPException
from bson import ObjectId
from app.database import get_database
from app.middleware.auth import require_admin
from app.schemas.invoice import CreateInvoiceRequest, InvoiceResponse
from app.models.invoice import new_invoice

router = APIRouter()


@router.get("")
async def list_invoices(admin: dict = Depends(require_admin)):
    db = get_database()
    cursor = db.invoices.find().sort("created_at", -1)
    items = []
    async for inv in cursor:
        items.append(InvoiceResponse(
            id=str(inv["_id"]),
            client_id=str(inv["client_id"]),
            project_id=inv.get("project_id"),
            stripe_invoice_id=inv.get("stripe_invoice_id", ""),
            amount_cents=inv["amount_cents"],
            status=inv["status"],
            due_date=inv["due_date"],
            line_items=inv["line_items"],
            created_at=inv["created_at"],
            token=inv.get("token", ""),
            paid_at=inv.get("paid_at"),
        ))
    return {"invoices": items}


@router.post("", status_code=201)
async def create_invoice_endpoint(body: CreateInvoiceRequest, admin: dict = Depends(require_admin)):
    db = get_database()
    client = await db.users.find_one({"_id": ObjectId(body.client_id), "role": "client"})
    if not client:
        raise HTTPException(status_code=404, detail="Client not found")
    total_cents = sum(item.amount_cents * item.quantity for item in body.line_items)
    line_items_dicts = [li.model_dump() for li in body.line_items]
    invoice = new_invoice(
        client_id=body.client_id,
        amount_cents=total_cents,
        line_items=line_items_dicts,
        due_date=body.due_date,
        project_id=body.project_id,
    )
    result = await db.invoices.insert_one(invoice)
    return {"id": str(result.inserted_id), "message": "Invoice created"}


@router.get("/{invoice_id}")
async def get_invoice(invoice_id: str, admin: dict = Depends(require_admin)):
    db = get_database()
    inv = await db.invoices.find_one({"_id": ObjectId(invoice_id)})
    if not inv:
        raise HTTPException(status_code=404, detail="Invoice not found")
    return InvoiceResponse(
        id=str(inv["_id"]),
        client_id=str(inv["client_id"]),
        project_id=inv.get("project_id"),
        stripe_invoice_id=inv.get("stripe_invoice_id", ""),
        amount_cents=inv["amount_cents"],
        status=inv["status"],
        due_date=inv["due_date"],
        line_items=inv["line_items"],
        created_at=inv["created_at"],
        paid_at=inv.get("paid_at"),
    )


@router.put("/{invoice_id}/status")
async def update_invoice_status(invoice_id: str, body: dict, admin: dict = Depends(require_admin)):
    db = get_database()
    new_status = body.get("status")
    valid_statuses = ["draft", "sent", "paid", "void"]
    if new_status not in valid_statuses:
        raise HTTPException(status_code=400, detail=f"Status must be one of: {valid_statuses}")
    update: dict = {"status": new_status}
    if new_status == "paid":
        update["paid_at"] = datetime.utcnow()
    result = await db.invoices.update_one({"_id": ObjectId(invoice_id)}, {"$set": update})
    if result.matched_count == 0:
        raise HTTPException(status_code=404, detail="Invoice not found")
    return {"message": f"Invoice marked as {new_status}"}

File: src/routes/admin/invoices.ts
import { Router, Response } from 'express';
import { ObjectId } from 'mongodb';
import { getDb } from '../../database';
import { requireAdmin, AuthRequest } from '../../middleware/auth';
import { newInvoice } from '../../models/invoice';

const router = Router();

router.get('/', requireAdmin, async (_req: AuthRequest, res: Response): Promise<void> => {
  const db = await getDb();
  const invoices = await db.collection('invoices').find().sort({ created_at: -1 }).toArray();
  res.json({ invoices: invoices.map((inv) => ({ id: inv._id.toString(), client_id: inv.client_id, project_id: inv.project_id ?? null, stripe_invoice_id: inv.stripe_invoice_id ?? '', amount_cents: inv.amount_cents, status: inv.status, due_date: inv.due_date, line_items: inv.line_items, created_at: inv.created_at, token: inv.token ?? '', paid_at: inv.paid_at ?? null })) });
});

router.post('/', requireAdmin, async (req: AuthRequest, res: Response): Promise<void> => {
  const { client_id, line_items, due_date, project_id } = req.body;
  if (!client_id || !line_items || !due_date) { res.status(400).json({ detail: 'client_id, line_items, and due_date required' }); return; }
  const db = await getDb();
  const client = await db.collection('users').findOne({ _id: new ObjectId(client_id), role: 'client' });
  if (!client) { res.status(404).json({ detail: 'Client not found' }); return; }
  const totalCents = line_items.reduce((sum: number, item: { amount_cents: number; quantity: number }) => sum + item.amount_cents * item.quantity, 0);
  const invoice = newInvoice(client_id, totalCents, line_items, new Date(due_date), project_id ?? null);
  const result = await db.collection('invoices').insertOne(invoice);
  res.status(201).json({ id: result.insertedId.toString(), message: 'Invoice created' });
});

router.get('/:invoiceId', requireAdmin, async (req: AuthRequest, res: Response): Promise<void> => {
  const db = await getDb();
  const inv = await db.collection('invoices').findOne({ _id: new ObjectId(req.params.invoiceId) });
  if (!inv) { res.status(404).json({ detail: 'Invoice not found' }); return; }
  res.json({ id: inv._id.toString(), client_id: inv.client_id, project_id: inv.project_id ?? null, stripe_invoice_id: inv.stripe_invoice_id ?? '', amount_cents: inv.amount_cents, status: inv.status, due_date: inv.due_date, line_items: inv.line_items, created_at: inv.created_at, token: inv.token ?? '', paid_at: inv.paid_at ?? null });
});

router.put('/:invoiceId/status', requireAdmin, async (req: AuthRequest, res: Response): Promise<void> => {
  const valid = ['draft', 'sent', 'paid', 'void'];
  const { status } = req.body;
  if (!valid.includes(status)) { res.status(400).json({ detail: `Status must be one of: ${valid.join(', ')}` }); return; }
  const db = await getDb();
  const update: Record<string, unknown> = { status };
  if (status === 'paid') update.paid_at = new Date();
  const result = await db.collection('invoices').updateOne({ _id: new ObjectId(req.params.invoiceId) }, { $set: update });
  if (!result.matchedCount) { res.status(404).json({ detail: 'Invoice not found' }); return; }
  res.json({ message: `Invoice marked as ${status}` });
});

export default router;
